File: scala/binary-search/src/main/scala/BinarySearch.scala
object BinarySearch {
  def find[T](items: Seq[T], item: T)(implicit orderer: T => Ordered[T]): Option[Int] =
    items match {
      case Nil => None
      case x :: Nil if x == item => Some(0)
      case _ :: Nil => None
      case _ =>
        val middle = items.length / 2
        items(middle) match {
          case x if x == item => Some(middle)
          case x if x > item => find(items.take(middle), item)
          case x if x < item => find(items.drop(middle), item).map(_ + middle)
        }
    }
}

File: kotlin/binary-search/src/main/kotlin/BinarySearch.kt
object BinarySearch {
    fun search(list: List<Int>, item: Int): Int {
        var (lo, hi) = Pair(0, list.size - 1)

        while (lo <= hi) {
            val mid = (lo + hi) / 2

            if (list[mid] == item) {
                return mid
            }

            if (list[mid] < item) {
                lo = mid + 1
            } else {
                hi = mid - 1
            }
        }

        throw NoSuchElementException()
    }
}
